### PR TITLE
Mean

### PR DESCRIPTION
**Pull request by OleNet**
### PR types

New features
### PR changes

OPs
### Describe

image

### CODE REVIEW

File: paddle/fluid/operators/mean_op_npu.cc [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    mean_tensor.mutable_data<T>(context.GetPlace());
+    std::vector<float> mean_vec;
+    mean_vec.push_back(1.0/static_cast<float>(IG->numel()));
+    framework::TensorFromVector(mean_vec,
+                                context.device_context(),
+                                &mean_tensor);
+
+    // means mul ones
+    Tensor mean_ma(grad->type());
+    mean_ma.Resize(framework::make_ddim(dout_dims));
```
Comment (liym27) [inline paddle/fluid/operators/mean_op_npu.cc]: `mean_ma.Resize(framework::make_ddim(dout_dims));`
->

`mean_ma.Resize(framework::make_ddim(IG->dims()));`

File: paddle/fluid/operators/mean_op_npu.cc [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    auto* out = ctx.Output<framework::LoDTensor>("Out");
+
+    std::vector<int> axes;
+
+    framework::NPUAttributeMap attr_input = {
+                  {"keep_dims", false},
+                  {"axes", axes}};
+
+    std::vector<int64_t> out_dims;
+    out_dims.push_back(1);
```
Comment (liym27) [inline paddle/fluid/operators/mean_op_npu.cc]: Remove line 33 - 34

File: paddle/fluid/operators/mean_op_npu.cc [hunk 3]
```diff
@@ -0,0 +1,123 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (liym27) [inline paddle/fluid/operators/mean_op_npu.cc]: 2016 -> 2021

File: paddle/fluid/operators/mean_op_npu.cc [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    auto IG = context.Output<Tensor>(framework::GradVarName("X"));
+    IG->mutable_data<T>(context.GetPlace());
+
+    // ones
+    Tensor ones(grad->type());
+    std::vector<int64_t> dout_dims;
+    for (auto i = 0; i < IG->dims().size(); ++i) {
+      dout_dims.push_back(IG->dims()[i]);
+    }
```
Comment (liym27) [inline paddle/fluid/operators/mean_op_npu.cc]: Remove line 73 - 76

File: paddle/fluid/operators/mean_op_npu.cc [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    mean_ma.Resize(framework::make_ddim(dout_dims));
+    mean_ma.mutable_data<T>(context.GetPlace());
+    auto runner_mul_1 = NpuOpRunner("Mul", {mean_tensor, ones}, {mean_ma}, {});
+    runner_mul_1.Run(stream);
+
+    // and mul grad
+    auto runner_mul_2 = NpuOpRunner("Mul", {mean_ma, *grad}, {*IG}, {});
+    runner_mul_2.Run(stream);
+  }
+};
```
Comment (liym27) [inline paddle/fluid/operators/mean_op_npu.cc:96]: Lines 91-100, why not use only one "Mul", which input are "mean_tensor" and "*grad", output is "*IG"?

Comment (pangyoki) [reply paddle/fluid/operators/mean_op_npu.cc:96]: `mean_tensor`和`*grad`的shape都是{1}，请看下是否需要加上一个`BroadcastToD`的操作，变成`*IG`的维度

**Review by pangyoki — COMMENTED** [review]
请上传`mean_grad` kernel 在NPU上执行的glog截图

**Review by liym27 — APPROVED** [review]
LGTM

**Pull request merged**